Change TCPStore world_size and is_master to be optional

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51809 Change TCPStore world_size and is_master to be optional**

Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

H-Huang committed:
Change TCPStore world_size and is_master to be optional

[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"




[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"




[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

cbalioglu commented on test/distributed/test_c10d.py:
```diff
@@ -81,6 +82,14 @@ def gpus_for_rank(world_size):
     return gpus_for_rank
 
+def create_client(i, addr, port, world_size=-1, timeout=timedelta(seconds=10)):
+    if world_size < 0:
```
nitpick: Do you need a conditional here? Passing `world_size=-1` should behave anyways as if it was not provided.

osalpekar commented on test/distributed/test_c10d.py:
```diff
@@ -354,6 +363,31 @@ def test_compare_set(self):
         self.assertEqual(b"new_value0", store.get("key0"))
 
+    def test_multi_worker_with_fixed_world_size(self):
+        addr = "localhost"
```
nit: Should we make this a standard string constant across the file, like "DEFAULT_HOSTNAME"?

osalpekar commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -677,9 +677,9 @@ pair, :meth:`~torch.distributed.store.get` to retrieve a key-value pair, etc.
-    is_master (bool): True when initializing the server store, False for client stores.
-    timeout (timedelta): Timeout used by the store during initialization and for methods such as :meth:`~torch.distributed.store.get` and :meth:`~torch.distributed.store.wait`.
+    world_size (int, optional): The total number of store users (number of clients + 1 for the server). Default is -1 (a negative value indicates an non-fixed number of store users).
+    is_master (bool, optional): True when initializing the server store, False for client stores. Default is False.
```
Even if this is optional, we must have one server store right, since none of the ranks would store any data if all are clients? Should we make that explicit in the docs?

H-Huang replied:
Yes, the client would just hang until it is able to establish a connection to the server. Good point, I added a note.

osalpekar commented on test/distributed/test_c10d.py:
```diff
@@ -354,6 +363,31 @@ def test_compare_set(self):
+            p.join()
+
+    def test_multi_worker_with_nonfixed_world_size(self):
+        addr = "localhost"
```
nit: same as above

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

H-Huang commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -66,7 +66,7 @@ class TCPStore : public Store {
       const std::string& masterAddr,
       PortType masterPort,
-      int numWorkers,
+      int numWorkers = -1,
```
Thanks! Added this.

H-Huang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -701,10 +701,10 @@ Example::
           py::arg("host_name"),
           py::arg("port"),
-          py::arg("world_size"),
+          py::arg("world_size") = -1,
```
Since this is through pybind, I was not able to create an optional None argument without the extra complexity of defining a separate class for this argument. Pybind supports using [std::optional](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#allow-prohibiting-none-arguments) but we are on C++ 14 which does not have std::optional so I decided to leave it as is. However, I did change the C++ constructor and I will add the waitWorkers param / getPort method in a separate PR.

cbalioglu replied:
Although not "Pythonic", fair enough. As a side note; pybind11's support for standard container types can be extended via a `type_caster<>`. If you have a type that is API compatible with an STL type (e.g. `c10::optional`), you can easily support it in your Python interop. See [here](https://docs.google.com/document/d/15Caw9eGSE949xljF59YZb2FPR2Z762Pc9GVZU4aVFh4) for more info.

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -416,7 +416,7 @@ TCPStore::TCPStore(
   storeSocket_ = tcputil::connect(
       tcpStoreAddr_, tcpStorePort_, /* wait= */ true, timeout_);
-  if (waitWorkers) {
+  if (numWorkers >= 0 && waitWorkers) {
```
Before the comparison, you should first make sure that `numWorkers` has a value.

```
if (numWorkers && *numWorkers >= 0 && waitWorkers) {
  ...
}
```

-or-

```
if (numWorkers.value_or(-1) >= 0 && waitWorkers) {
  ...
}
```

cbalioglu replied:
You should also make sure that all other uses of `numWorkers` in the code have a similar check.

H-Huang replied:
Thanks!

cbalioglu commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -66,7 +66,7 @@ class TCPStore : public Store {
       const std::string& masterAddr,
       PortType masterPort,
-      int numWorkers,
+      c10::optional<int> numWorkers = -1,
```
The default value should be `c10::nullopt_t` instead of `-1`.

H-Huang committed:
Update on "Change TCPStore world_size and is_master to be optional"


Changes to TCPStore which will make world_size and is_master optional parameters for initialization.

API before change:
```python
# arguments: host_name, port, world_size, is_master, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, 2, True)
client_store = dist.TCPStore("127.0.0.1", 0, 2, False)
```

API after change:
```python
# arguments: host_name, port, world_size=-1, is_master=False, timeout=300s
server_store = dist.TCPStore("127.0.0.1", 0, is_master=True)
client_store = dist.TCPStore("127.0.0.1", 0)
```

Differential Revision: [D26461770](https://our.internmc.facebook.com/intern/diff/D26461770)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -66,7 +66,7 @@ class TCPStore : public Store {
       const std::string& masterAddr,
       PortType masterPort,
-      int numWorkers,
+      c10::optional<int> numWorkers = c10::nullopt_t(-1),
```
nitpick: Should be `c10::optional<int> numWorkers = c10::nullopt;`

cbalioglu reviewed (approved):
Just one minor comment. No need to review it though.
